Document string context (#8595)
NixOS/nix: doc/manual/src/language/string-context.md

> # String context

Reviewer: This needs a warning telling that it's an advanced thing that people probably don't need to know unless they really do

Strings in the Nix language are not just a sequences of characters like strings in other languages.
They are actually pairs of sequences of characters and what is known as a *string context*.
The string context is an (unordered) set of *string context elements*.

> 

Reviewer: Maybe add a quick sentence about the goal of string contexts.

String context elements come in 3 forms:

- [*constant*]{#string-context-element-constant}
- [*output*]{#string-context-element-output}
- [*derivation deep*]{#string-context-element-derivation-deep}

*Constant* and *output* string contexts elements are just
[derived paths](@docroot@/glossary.md#gloss-derived-path);
those are just the names of the two kinds of derived path.
See the documentation on derived paths for further details.

> *derivation deep* is far more obscure.
> It is intended to be used with the

Reviewer: ```suggestion
*derivation deep* is an advanced feature intended to be used with the
```

When more granular information is needed, [`builtins.getContext`] can be used.
It creates an [attribute set] representing the string context, which can be inspected as usual.

[`builtins.hasContext`]: ./builtins.md#builtins-hasContext
[`builtins.getContext`]: ./builtins.md#builtins-getContext
[attribute set]: ./values.md#attribute-set

## Clearing string contexts

[`buitins.unsafeDiscardStringContext`](./builtins.md#) will make a copy of a string, but with an empty string context.
The returned string can be used in more ways, e.g. by operators that require the string context to be empty.
The requirement to explicitly discard the string context in such use cases helps ensure that string context elements are not lost by mistake.
The "unsafe" marker is only there to remind that Nix normally guarantees that dependencies are tracked, whereas the returned string has lost them.

## Constructing string contexts

[`builtins.appendContext`] will create a copy of a string, but with additional string context elements.
The context is specified explicitly by an [attribute set] in the format that [`builtins.hasContext`] produces.
A string with arbitrary contexts can be made like this:

> 1. Create a string with the desired string context elements.

Author: ```suggestion
1. Create a string with the desired string context elements.
   (The contents of the string do not matter.)
```

# String context

> **Note**
>
> This is an advanced topic.
> The Nix language is designed to be used without the programmer consciously dealing with string contexts or even knowing what they are.

A string in the Nix language is not just a sequence of characters like strings in other languages.
It is actually a pair of a sequence of characters and a *string context*.
The string context is an (unordered) set of *string context elements*.

The purpose of string contexts is to collect non-string values attached to strings via
[string concatenation](./operators.md#string-concatenation),
[string interpolation](./string-interpolation.md),
and similar operations.
The idea is that a user can combine together values to create a build recipe without manually keeping track of where the "ingredients" come from, and then the Nix language does that bookkeeping implicitly to come up with the right derivation inputs.

> > In line with this goal, string contexts are *not* explicitly manipulated in idiomatic Nix code.

Reviewer 2: ```suggestion
> In line with this goal, string contexts are *not* explicitly manipulated in idiomatic Nix language code.
```

Let's be clear on the distinction!

Reviewer 2: ```suggestion
> **Note**
>
> String contexts are *not* explicitly manipulated in idiomatic Nix language code.
```

> > Strings with non-empty contexts are only concatenated and eventually passed to `builtins.derivation`.
> > Plain strings with empty contexts are the only ones to be inspected, e.g. using comparison with `==`.

Reviewer 2: ```suggestion
Strings with non-empty contexts are only concatenated and eventually passed to [`builtins.derivation`](./derivations.md).
Plain strings with empty contexts are the only ones to be inspected, e.g. using comparison with [the equality operator (`==`)](./operators#equality).
```

No reason for this to be part of the highlight.

3. Combine it with a base string and repeated [`builtins.appendContext`] calls.

The remainder of this section will focus on step 1: making strings with individual string context elements on which to apply `builtins.getContext`.

[`builtins.appendContext`]: ./builtins.md#builtins-appendContext

### Constant string context elements

A constant string context element is just a constant [deriving path];
a constant deriving path is just a [store path].
We therefore want to use [`builtins.storePath`] to create a string with a single constant string context element:

> **Example**
>
> ```nix
> builtins.getContext (builtins.storePath "/nix/store/wkhdf9jinag5750mqlax6z2zbwhqb76n-hello-2.10")
> ```
> evaluates to
> ```nix
> {
>   "/nix/store/wkhdf9jinag5750mqlax6z2zbwhqb76n-hello-2.10" = {
>     path = true;
>   };
> }
> ```

[deriving path]: @docroot@/glossary.md#gloss-deriving-path
[store path]: @docroot@/glossary.md#gloss-store-path
[`builtins.storePath`]: ./builtins.md#builtins-storePath

### Output string context elements

> **Example**
>
> This is best illustrated with a built-in function that is still experimental: [`builtins.ouputOf`].

> > This example will *not* work the stable Nix!

Reviewer 2: ```suggestion
> This example will *not* work with stable Nix!
```

[`exportReferencesGraph`](./advanced-attributes.html#adv-attr-exportReferencesGraph)
advanced derivation feature.
A *derivation deep* string context element is a derivation path, and refers to both its outputs and the entire build closure of that derivation:
all its outputs, all the other derivations the given derivation depends on, and all their outputs too.

## Inspecting string contexts

Most basically, [`builtins.hasContext`] will whether a string as a non-empty context.

When more granular information than merely whether the string context is empty is needed, [`builtins.getContext`] can be used.
It creates an [attribute set] representing the string context; the attribute set can be inspected normally.

[`builtins.hasContext`]: ./builtins.md#builtins-hasContext
[`builtins.getContext`]: ./builtins.md#builtins-getContext
[attribute set]: ./values.md#attribute-set

## Clearing string contexts

[`buitins.unsafeDiscardStringContext`](./builtins.md#) will make a copy of another string, but with an empty string context.
The new string can be inspected in more ways, e.g. by operators that require that the string context is empty.

> Explicitly discarding the string context and then expecting it helps ensure that string context elements are not lost by mistake.

Author: ```suggestion
The requirement to explicitly discard the string context in such use cases, helps ensure that string context elements are not lost by mistake.
The "unsafe" marker is only there to remind that Nix normally guarantees that dependencies are tracked, whereas the returned string has lost them.
```

1. Create strings with the string context elements we want
2. Dump their contexts with [`builtins.getContext`]
3. Combine them together with a base string and repeated [`builtins.appendContext`] calls.

The remainder of this section will focus on step 1: making strings with individual string context elements on which to apply `builtins.getContext`.

[`builtins.appendContext`]: ./builtins.md#builtins-appendContext

## Constant string context elements

A constant string context element is just a constant [derived path];
a constant derived path is just a [store path].
We therefore want to use [`builtins.storePath`] to create a string with a single constant string context element:

> **Example**
>
> ```nix
> builtins.getContext (builtins.storePath "/nix/store/wkhdf9jinag5750mqlax6z2zbwhqb76n-hello-2.10")
> ```
> evaluates to
> ```nix
> {
>   "/nix/store/wkhdf9jinag5750mqlax6z2zbwhqb76n-hello-2.10" = {
>     path = true;
>   };
> }
> ```

[derived path]: @docroot@/glossary.md#gloss-derived-path
[store path]: @docroot@/glossary.md#gloss-store-path
[`builtins.storePath`]: ./builtins.md#builtins-storePath

## Output string context elements

> The best way to illustrate this with a builtin function that is still experimental: [`builtins.ouputOf`].

Reviewer 3: ```suggestion
This is best illustrated with a builtin function that is still experimental: [`builtins.ouputOf`].
```

(missing "is")

This example will *not* work the stable Nix!

> **Example**
>
> ```nix
> builtins.getContext
>   (builtins.outputOf
>     (builtins.storePath "/nix/store/fvchh9cvcr7kdla6n860hshchsba305w-hello-2.12.drv")
>     "out")
> ```
> evaluates to
> ```nix
> {
>   "/nix/store/fvchh9cvcr7kdla6n860hshchsba305w-hello-2.12.drv" = {
>     outputs = [ "out" ];
>   };
> }
> ```

[`builtins.outputOf`]: ./builtins.md#builtins-outputOf

## "Derivation deep" string context elements

> These best way to illustrate this is with [`builtins.addDrvOutputDependencies`].

Reviewer 3: ```suggestion
The best way to illustrate this is with [`builtins.addDrvOutputDependencies`].
```

> We take a regular constant string context element pointing to a derivation, and transform it unto a "Derivation deep" string context element.

Reviewer 3: ```suggestion
We take a regular constant string context element pointing to a derivation, and transform it into a "Derivation deep" string context element.
```

> **Example**
>
> ```nix
> builtins.getContext
>   (builtins.addDrvOutputDependencies
>     (builtins.storePath "/nix/store/fvchh9cvcr7kdla6n860hshchsba305w-hello-2.12.drv"))
> ```
> evaluates to
> ```nix
> {
>   "/nix/store/fvchh9cvcr7kdla6n860hshchsba305w-hello-2.12.drv" = {
>     allOutputs = true;
>   };
> }
> ```

> [`builtins.unsafeDiscardOutputDependency`] does this the opposite of [`builtins.addDrvOutputDependencies`], but is not prefered because it "weakens" rather than "strengens" the string context.

Reviewer 3: ```suggestion
[`builtins.unsafeDiscardOutputDependency`] does the opposite of [`builtins.addDrvOutputDependencies`], but is not preferred because it "weakens" rather than "strengthens" the string context.
```

Reviewer 3: ```suggestion
[`builtins.unsafeDiscardOutputDependency`] does this the opposite of [`builtins.addDrvOutputDependencies`], but is marked unsafe because it "forgets" the dependencies on the outputs, whereas Nix normally tracks all dependencies consistently.
```

probably don't need the next sentence if you apply this.

# String context

> **Note**
>
> This is an advanced topic.
> The Nix language is designed to be used without the programmer knowing string contexts are.

Strings in the Nix language are not just a sequences of characters like strings in other languages.
They are actually pairs of sequences of characters and what is known as a *string context*.
The string context is an (unordered) set of *string context elements*.

> The goal of string contexts are to collect non-string values combined with strings via

Author: Yes it is about that, your suggestion looks good to me :).

[string concatenation](./operators.md#string-concatenation),
[string interpolation](./string-interpolation.md),

> and similar operations.

Author: interpolation is sugar for concatentation. (Our concatenation also implies `toString` as needed).

> The idea is that a user can combine together values to create a build recipe without manually keeping track of where the "ingredients" come from, and then the Nix language does that bookkeeping implicitly to come up with the right derivation inputs.

Author: I say "values" because arguably the things we are concatentaing aren't really strings proper. The string context in purpose a way to concatenate strings and *non strings* together --- by expanding the definition of "string" to accommodate these other things.

> The idea is that a user can combine together values to create a build recipe without manually keeping track of where the "ingredients" come from, and then the Nix language does that bookkeeping implicitly to come up with the right derivation inputs.

Reviewer 2: ```suggestion
The idea is that a user can combine values into build instructions for derivations without manually keeping track of where they come from.
Then the Nix language implicitly does that bookkeeping to efficiently obtain the closure of derivation inputs.
```

> In line with this goal, string contexts are *not* explicitly manipulated in idiomatic Nix code.

> > Strings with non-empty context are only concatenated and eventually passed to `builtins.derivation`.

Author: ```suggestion
> Strings with non-empty contexts are only concatenated and eventually passed to `builtins.derivation`.
```

> > "Regular" strings an empty context are the only ones inspected, e.g. compared with `==`.

Author: ```suggestion
> Plain strings with an empty context are the only ones to be inspected, e.g. using comparison with `==`.
```

String context elements come in 3 forms:

- [*constant*]{#string-context-element-constant}
- [*output*]{#string-context-element-output}
- [*derivation deep*]{#string-context-element-derivation-deep}

*Constant* and *output* string contexts elements are just
[deriving paths](@docroot@/glossary.md#gloss-deriving-path);
those are just the names of the two kinds of deriving path.
See the documentation on deriving paths for further details.

*derivation deep* is an advanced feature intended to be used with the
[`exportReferencesGraph`](./advanced-attributes.html#adv-attr-exportReferencesGraph)
advanced derivation feature.
A *derivation deep* string context element is a derivation path, and refers to both its outputs and the entire build closure of that derivation:

> all its outputs, all the other derivations the given derivation depends on, and all their outputs too.

Author: I feel weird not given each item its own bullet point, however. Especially since we removed the `3`.

We could however really embrace the subgroup and say that a string context is either a derived path or a derivation deep, and let the user read about derived paths separately. Then it really would be 2 kinds, with 2 kinds within the first kind.

> - [*constant*]{#string-context-element-constant}
> - [*output*]{#string-context-element-output}
> - [*derivation deep*]{#string-context-element-derivation-deep}
> 
> *Constant* and *output* string contexts elements are just
> [deriving paths](@docroot@/glossary.md#gloss-deriving-path);
> those are just the names of the two kinds of deriving path.
> See the documentation on deriving paths for further details.
> 
> *derivation deep* is an advanced feature intended to be used with the
> [`exportReferencesGraph`](./advanced-attributes.html#adv-attr-exportReferencesGraph)
> advanced derivation feature.
> A *derivation deep* string context element is a derivation path, and refers to both its outputs and the entire build closure of that derivation:
> all its outputs, all the other derivations the given derivation depends on, and all their outputs too.

Reviewer 2: ```suggestion
- [deriving path]{#string-context-element-derived-path}

  A string context element of this type is a [deriving path](@docroot@/glossary.md#gloss-deriving-path).
  They can be one of the varieties *constant* or *output*, which correspond to the types of deriving path.
  
  <!-- put the two sections with examples here as a list with anchors on the items -->

- [*derivation deep*]{#string-context-element-derivation-deep}

  *derivation deep* is an advanced feature intended to be used with the
  [`exportReferencesGraph` derivation attribute](./advanced-attributes.html#adv-attr-exportReferencesGraph).

  A *derivation deep* string context element is a derivation path, and refers to both its outputs and the entire build closure of that derivation:
  all its outputs, all the other derivations the given derivation depends on, and all the outputs of those.
```

[`builtins.getContext`]: ./builtins.md#builtins-getContext
[attribute set]: ./values.md#attribute-set

## Clearing string contexts

[`buitins.unsafeDiscardStringContext`](./builtins.md#) will make a copy of a string, but with an empty string context.
The returned string can be used in more ways, e.g. by operators that require that the string context is empty.
The requirement to explicitly discard the string context in such use cases, helps ensure that string context elements are not lost by mistake.
The "unsafe" marker is only there to remind that Nix normally guarantees that dependencies are tracked, whereas the returned string has lost them.

## Constructing string contexts

[`builtins.appendContext`] will create copy of a string but with additional string context elements.
The context is specified explicitly by an [attribute set] in the format that [`builtins.hasContext`] produces.
We can create strings with arbitrary contexts in 3 steps:

1. Create strings with the string context elements we want
2. Dump their contexts with [`builtins.getContext`]
3. Combine them together with a base string and repeated [`builtins.appendContext`] calls.

> The remainder of this section will focus on step 1: making strings with individual string context elements on which to apply `builtins.getContext`.
> 

Author: I think it needs some transition, but I think I can make it less "order-dependent". How about this?

```suggestion
Steps (2) and (3) are the same regardless of the type of element we are trying to include in our final string context.
Step (1) however is specific to the type of string context in question.
Now we describe step (1) for each type of string context element.
```

> 

Reviewer 2: Let's instead move the subsections into the list with the two string context types, as suggested in the last review.

[`builtins.appendContext`]: ./builtins.md#builtins-appendContext

## Constant string context elements

A constant string context element is just a constant [deriving path];

> a constant deriving path is just a [store path].

Author: I am trying to distinguish the proverbial "sum type data constructor" vs the single field.

"constant deriving path" in *colloquial* speech is just a synonymy for "store path", but it would be mean to create redundant jargon words for the same constant. Formally "constant" and "output", describe the two cases of "deriving path". I guess we can think of these data structures as adjectives:

>  A deriving path is either "constant" [adj] in which case it is a store path, or "output" (forced adjective, awk) in which case it a pair of a store path an an output name.

> a constant deriving path is just a [store path].

Author: My using "just" was the most abbreviated and unclear (and subconcious) way of dealing

> a constant deriving path is just a [store path].

Reviewer 2: Does the suggestion still make sense then?

> ```nix
> {
>   "/nix/store/fvchh9cvcr7kdla6n860hshchsba305w-hello-2.12.drv" = {
>     outputs = [ "out" ];
>   };
> }
> ```

[`builtins.outputOf`]: ./builtins.md#builtins-outputOf

## "Derivation deep" string context elements

The best way to illustrate this is with [`builtins.addDrvOutputDependencies`].
We take a regular constant string context element pointing to a derivation, and transform it into a "Derivation deep" string context element.

> **Example**
>
> ```nix
> builtins.getContext
>   (builtins.addDrvOutputDependencies
>     (builtins.storePath "/nix/store/fvchh9cvcr7kdla6n860hshchsba305w-hello-2.12.drv"))
> ```
> evaluates to
> ```nix
> {
>   "/nix/store/fvchh9cvcr7kdla6n860hshchsba305w-hello-2.12.drv" = {
>     allOutputs = true;
>   };
> }
> ```

[`builtins.unsafeDiscardOutputDependency`] does this the opposite of [`builtins.addDrvOutputDependencies`], but is marked unsafe because it "forgets" the dependencies on the outputs, whereas Nix normally tracks all dependencies consistently.
[`builtins.addDrvOutputDependencies`] in contrast is safe because "derivation deep" string context element always refers to the underlying derivation (among many more things);

> replacing a constant string context element with a "derivation deep" element is a safe operation that just enlargens the string context without forgetting anything.

Author: Moved it.
